File: gcs/windows/src/BumblebeeGcs/Services/WifiHandoff.cs
// Copyright 2026 FutureLab
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System.Security;
using ManagedNativeWifi;
using Microsoft.UI.Dispatching;
using Windows.Devices.Geolocation;

namespace BumblebeeGcs.Services;

/// <summary>
/// One-tap передача WiFi: найти setup-хотспот дрона в эфире, подключиться к нему,
/// отправить дрону креды домашней сети, дождаться его возврата в общую сеть.
/// Порт WifiHandoff.swift (CoreWLAN → wlanapi через ManagedNativeWifi,
/// CoreLocation → Geolocator: Windows 11 требует геолокацию для скана WiFi).
/// Все публичные члены — UI-поток.
/// </summary>
public sealed class WifiHandoff
{
    /// <summary>Общий PSK setup-хотспота дрона (SSID "Bumblebee-XXXX").</summary>
    public const string HotspotPsk = "12345678";
    /// <summary>Префикс каждого setup-хотспота.</summary>
    public const string HotspotPrefix = "Bumblebee-";
    /// <summary>Дрон в режиме AP (NetworkManager shared) — шлюз по этому адресу.</summary>
    public const string HotspotGatewayHost = "10.42.0.1";

    public enum PhaseKind
    {
        Idle,
        NeedLocationPermission,
        Scanning,
        Picking,           // хотспоты показаны, ждём выбора
        JoiningHotspot,
        PushingCreds,
        WaitingForDrone,
        Done,              // Arg = хост дрона
        Failed,            // Arg = сообщение
    }

    public PhaseKind Phase { get; private set; } = PhaseKind.Idle;
    /// <summary>SSID для JoiningHotspot, хост для Done, сообщение для Failed.</summary>
    public string PhaseArg { get; private set; } = "";

    public bool IsBusy => Phase is PhaseKind.Scanning or PhaseKind.JoiningHotspot
        or PhaseKind.PushingCreds or PhaseKind.WaitingForDrone;

    public sealed record Hotspot(string Ssid, int Rssi);

    public List<Hotspot> Hotspots { get; private set; } = new();
    /// <summary>Текущий SSID этого ПК — сеть, которую отдаём дрону.</summary>
    public string CurrentSsid { get; private set; } = "";
    /// <summary>Система окончательно запретила геолокацию — шлём в Настройки.</summary>
    public bool LocationBlocked { get; private set; }

    public event Action? Changed;

    private readonly DispatcherQueue _dq;

    public WifiHandoff(DispatcherQueue dq) => _dq = dq;

    private void SetPhase(PhaseKind phase, string arg = "")
    {
        Phase = phase;
        PhaseArg = arg;
        Changed?.Invoke();
    }

    public void Reset() => SetPhase(PhaseKind.Idle);

    public void RefreshCurrentSsid()
    {
        try { CurrentSsid = NativeWifi.EnumerateConnectedNetworkSsids().FirstOrDefault()?.ToString() ?? ""; }
        catch { CurrentSsid = ""; }
        Changed?.Invoke();
    }

    // MARK: - Location (нужна для скана WiFi на Windows 11)

    /// <summary>
    /// Разово запросить доступ к геолокации при старте, чтобы разрешение было
    /// выяснено до настройки дрона (аналог requestLocation на маке).
    /// </summary>
    public async Task RequestLocationAsync()
    {
        try
        {
            var access = await Geolocator.RequestAccessAsync();
            LocationBlocked = access == GeolocationAccessStatus.Denied;
        }
        catch { }
        Changed?.Invoke();
    }

    private static Guid? PrimaryInterfaceId()
    {
        try { return NativeWifi.EnumerateInterfaces().FirstOrDefault()?.Id; }
        catch { return null; }
    }

    // MARK: - Scan for drone hotspots

    public async Task ScanHotspotsAsync()
    {
        if (PrimaryInterfaceId() is null)
        {
            SetPhase(PhaseKind.Failed, "No WiFi interface on this PC");
            return;
        }
        try
        {
            var access = await Geolocator.RequestAccessAsync();
            LocationBlocked = access == GeolocationAccessStatus.Denied;
            if (access != GeolocationAccessStatus.Allowed)
            {
                SetPhase(PhaseKind.NeedLocationPermission);
                return;
            }
        }
        catch { }

        SetPhase(PhaseKind.Scanning);
        RefreshCurrentSsid();
        try
        {
            var list = await Task.Run(async () =>
            {
                try { await NativeWifi.ScanNetworksAsync(TimeSpan.FromSeconds(6)); } catch { }
                return NativeWifi.EnumerateBssNetworks()
                    .Where(n => n.Ssid.ToString().StartsWith(HotspotPrefix, StringComparison.Ordinal))
                    .GroupBy(n => n.Ssid.ToString())
                    .Select(g => new Hotspot(g.Key, g.Max(n => n.SignalStrength)))
                    .OrderByDescending(h => h.Rssi)
                    .ToList();
            });
            Hotspots = list;
            SetPhase(PhaseKind.Picking);
        }
        catch (Exception ex)
        {
            SetPhase(PhaseKind.Failed, $"WiFi scan failed: {ex.Message}");
        }
    }

    // MARK: - Full hand-off

    /// <summary>
    /// Подключиться к hotspotSsid, отправить (targetSsid, targetPassword) дрону,
    /// вернуть ПК в свою сеть и дождаться дрона. onDrone(host) при успехе.
    /// </summary>
    public async Task ProvisionAsync(string hotspotSsid, string targetSsid, string targetPassword,
                                     Action<string> onDrone)
    {
        if (PrimaryInterfaceId() is not Guid iface)
        {
            SetPhase(PhaseKind.Failed, "No WiFi interface on this PC");
            return;
        }

        // 1) Подключиться к setup-хотспоту дрона. Первая попытка может упасть,
        //    пока адаптер переключается — ретраим до 4 раз со свежим сканом.
        SetPhase(PhaseKind.JoiningHotspot, hotspotSsid);
        var lastErr = "";
        var joined = false;
        for (var attempt = 1; attempt <= 4 && !joined; attempt++)
        {
            try
            {
                joined = await Task.Run(async () =>
                {
                    var inRange = NativeWifi.EnumerateBssNetworks().Any(n => n.Ssid.ToString() == hotspotSsid);
                    if (!inRange)
                    {
                        try { await NativeWifi.ScanNetworksAsync(TimeSpan.FromSeconds(5)); } catch { }
                        inRange = NativeWifi.EnumerateBssNetworks().Any(n => n.Ssid.ToString() == hotspotSsid);
                    }
                    if (!inRange) throw new InvalidOperationException("hotspot no longer in range");
                    NativeWifi.SetProfile(iface, ProfileType.AllUser,
                        Wpa2PskProfileXml(hotspotSsid, HotspotPsk), null, overwrite: true);
                    return await NativeWifi.ConnectNetworkAsync(iface, hotspotSsid, BssType.Infrastructure,
                        TimeSpan.FromSeconds(15));
                });
                if (!joined) lastErr = "association failed";
            }
            catch (Exception ex)
            {
                lastErr = ex.Message;
            }
            if (!joined && attempt < 4) await Task.Delay(2000);
        }
        if (!joined)
        {
            SetPhase(PhaseKind.Failed, $"Couldn't join {hotspotSsid}: {lastErr}");
            return;
        }

        // 2) Дождаться бэкенда дрона на шлюзе хотспота.
        SetPhase(PhaseKind.PushingCreds);
        var api = new DroneApi { Host = HotspotGatewayHost };
        if (!await WaitForBackendAsync(api))
        {
            SetPhase(PhaseKind.Failed, $"Drone didn't respond on its hotspot ({HotspotGatewayHost})");
            await RejoinAsync(iface, targetSsid);
            return;
        }

        // 3) Отправить дрону WiFi этого ПК и применить (без авторизации).
        try
        {
            await api.SaveNetworkAsync(targetSsid, targetPassword);
            await api.ApplyNowAsync(targetSsid);
        }
        catch (Exception ex)
        {
            SetPhase(PhaseKind.Failed, $"Couldn't set WiFi on the drone: {ex.Message}");
            await RejoinAsync(iface, targetSsid);
            return;
        }

        // 4) Дрон гасит хотспот и уходит на targetSsid; возвращаем ПК в его сеть
        //    (профиль уже сохранён в Windows, раз мы в ней сидели).
        SetPhase(PhaseKind.WaitingForDrone);
        await RejoinAsync(iface, targetSsid);

        // 5) Ждём появления дрона в общей сети через DNS-SD.
        if (await WaitForDroneAsync() is string host)
        {
            SetPhase(PhaseKind.Done, host);
            onDrone(host);
        }
        else
        {
            SetPhase(PhaseKind.Failed, "Drone didn't come back online — check the WiFi password and retry");
        }
    }

    // MARK: - Helpers

    private static async Task RejoinAsync(Guid iface, string homeSsid)
    {
        if (string.IsNullOrEmpty(homeSsid)) return;
        try
        {
            await Task.Run(() =>
                NativeWifi.ConnectNetworkAsync(iface, homeSsid, BssType.Infrastructure, TimeSpan.FromSeconds(20)));
        }
        catch { }
    }

    /// <summary>Поллим бэкенд дрона, пока не ответит (или таймаут).</summary>
    private static async Task<bool> WaitForBackendAsync(DroneApi api, double timeoutSeconds = 25)
    {
        var deadline = DateTime.UtcNow.AddSeconds(timeoutSeconds);
        while (DateTime.UtcNow < deadline)
        {
            try
            {
                await api.WifiStatusAsync();
                return true;
            }
            catch { }
            await Task.Delay(1000);
        }
        return false;
    }

    /// <summary>Ждём любой дрон по DNS-SD, возвращаем его хост/IP.</summary>
    private static async Task<string?> WaitForDroneAsync(double timeoutSeconds = 60)
    {
        var tcs = new TaskCompletionSource<string?>(TaskCreationOptions.RunContinuationsAsynchronously);
        using var browser = new DnssdBrowser((ip, _) => tcs.TrySetResult(ip));
        browser.Start();
        var done = await Task.WhenAny(tcs.Task, Task.Delay(TimeSpan.FromSeconds(timeoutSeconds)));
        return done == tcs.Task ? tcs.Task.Result : null;
    }

    /// <summary>Профиль WPA2-PSK для временного подключения к хотспоту.</summary>
    internal static string Wpa2PskProfileXml(string ssid, string psk)
    {
        var name = SecurityElement.Escape(ssid);
        var key = SecurityElement.Escape(psk);
        return $"""
            <?xml version="1.0"?>
            <WLANProfile xmlns="http://www.microsoft.com/networking/WLAN/profile/v1">
                <name>{name}</name>
                <SSIDConfig><SSID><name>{name}</name></SSID></SSIDConfig>
                <connectionType>ESS</connectionType>
                <connectionMode>manual</connectionMode>
                <MSM>
                    <security>
                        <authEncryption>
                            <authentication>WPA2PSK</authentication>
                            <encryption>AES</encryption>
                            <useOneX>false</useOneX>
                        </authEncryption>
                        <sharedKey>
                            <keyType>passPhrase</keyType>
                            <protected>false</protected>
                            <keyMaterial>{key}</keyMaterial>
                        </sharedKey>
                    </security>
                </MSM>
            </WLANProfile>
            """;
    }
}

File: gcs/macos/Sources/GCSApp/WifiHandoff.swift
// Copyright 2026 FutureLab
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import Foundation
@preconcurrency import CoreWLAN
import CoreLocation

/// Fleet default PSK for a drone's setup hotspot (SSID "Bumblebee-XXXX").
/// The hotspot is raised by the drone's wifi-watchdog when it has no known WiFi.
let kHotspotPSK = "12345678"
/// Prefix every drone setup hotspot advertises.
let kHotspotPrefix = "Bumblebee-"
/// A drone in NetworkManager "shared" (AP) mode is the gateway at this address.
let kHotspotGatewayHost = "10.42.0.1"

/// One-tap WiFi hand-off: find a drone's setup hotspot over the air, join it,
/// push the Mac's own WiFi credentials to the drone, then wait for the drone to
/// come back on the shared network — all via CoreWLAN plus the drone's existing
/// `/api/wifi/*` backend. No System Keychain access (the target password is
/// entered once in-app), so macOS shows no admin prompt.
@MainActor
final class WifiHandoff: NSObject, ObservableObject, CLLocationManagerDelegate {
    enum Phase: Equatable {
        case idle
        case needLocationPermission
        case scanning
        case picking                 // hotspots listed for selection
        case joiningHotspot(String)
        case pushingCreds
        case waitingForDrone
        case done(String)            // resolved drone host
        case failed(String)

        var isBusy: Bool {
            switch self {
            case .scanning, .joiningHotspot, .pushingCreds, .waitingForDrone: return true
            default: return false
            }
        }
    }

    struct Hotspot: Identifiable, Equatable {
        let id: String       // ssid
        var ssid: String { id }
        var rssi: Int
    }

    @Published private(set) var phase: Phase = .idle
    @Published private(set) var hotspots: [Hotspot] = []
    /// The Mac's current WiFi SSID — the network we hand to the drone.
    @Published private(set) var currentSSID: String = ""

    private let loc = CLLocationManager()
    private var networksBySSID: [String: CWNetwork] = [:]

    override init() {
        super.init()
        loc.delegate = self
    }

    private var interface: CWInterface? { CWWiFiClient.shared().interface() }

    func reset() { phase = .idle }

    func refreshCurrentSSID() { currentSSID = interface?.ssid() ?? "" }

    // MARK: - Location (required for WiFi scan / SSID on macOS 14+)

    private func locationAuthorized() -> Bool {
        switch loc.authorizationStatus {
        case .authorizedAlways, .authorized, .authorizedWhenInUse: return true
        default: return false
        }
    }

    /// True once the system has made a decision we can't change in-app (denied
    /// or restricted) — the UI then points at System Settings instead of a
    /// prompt that will never appear again.
    var locationBlocked: Bool {
        loc.authorizationStatus == .denied || loc.authorizationStatus == .restricted
    }

    /// Proactively raise the system Location prompt if the user hasn't decided
    /// yet. Called at app launch so the permission is settled before setup.
    func requestLocation() {
        if loc.authorizationStatus == .notDetermined {
            loc.requestWhenInUseAuthorization()
        }
    }

    nonisolated func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        Task { @MainActor in
            if self.locationAuthorized(), self.phase == .needLocationPermission {
                await self.scanHotspots()
            }
        }
    }

    // MARK: - Scan for drone hotspots

    func scanHotspots() async {
        guard let iface = interface else { phase = .failed("No WiFi interface on this Mac"); return }
        guard locationAuthorized() else {
            phase = .needLocationPermission
            loc.requestWhenInUseAuthorization()
            return
        }
        phase = .scanning
        refreshCurrentSSID()
        do {
            let nets = try await Self.scan(iface)
            var map: [String: CWNetwork] = [:]
            var list: [Hotspot] = []
            for n in nets {
                guard let s = n.ssid, s.hasPrefix(kHotspotPrefix) else { continue }
                map[s] = n
                list.append(Hotspot(id: s, rssi: n.rssiValue))
            }
            networksBySSID = map
            hotspots = list.sorted { $0.rssi > $1.rssi }
            phase = .picking
        } catch {
            phase = .failed("WiFi scan failed: \(error.localizedDescription)")
        }
    }

    // MARK: - Full hand-off

    /// Join `hotspotSSID`, push (`targetSSID`,`targetPassword`) to the drone,
    /// let the Mac rejoin its network, then wait for the drone. Calls
    /// `onDrone(host)` with the resolved drone address on success.
    func provision(hotspotSSID: String,
                   targetSSID: String,
                   targetPassword: String,
                   onDrone: @escaping (String) -> Void) async {
        guard let iface = interface else {
            phase = .failed("No WiFi interface on this Mac"); return
        }
        // 1) Join the drone's setup hotspot. CoreWLAN's associate can throw a
        //    transient error ("tmpErr") when the interface is busy switching —
        //    retry a few times, re-fetching a fresh CWNetwork each attempt.
        phase = .joiningHotspot(hotspotSSID)
        var lastErr = ""
        var joined = false
        for attempt in 1...4 {
            var net = networksBySSID[hotspotSSID]
            if net == nil {
                net = (try? await Self.scan(iface))?.first(where: { $0.ssid == hotspotSSID })
            }
            guard let net else { lastErr = "hotspot no longer in range"; break }
            do { try iface.associate(to: net, password: kHotspotPSK); joined = true; break }
            catch {
                lastErr = error.localizedDescription
                if attempt < 4 { try? await Task.sleep(nanoseconds: 2_000_000_000) }
            }
        }
        guard joined else {
            phase = .failed("Couldn't join \(hotspotSSID): \(lastErr)"); return
        }

        // 2) Wait for the drone backend on the hotspot gateway.
        phase = .pushingCreds
        let api = DroneAPI()
        api.host = kHotspotGatewayHost
        guard await Self.waitForBackend(api) else {
            phase = .failed("Drone didn't respond on its hotspot (10.42.0.1)"); return
        }
        // 3) Push the Mac's WiFi and apply it on the drone (no auth required).
        do {
            try await api.saveNetwork(ssid: targetSSID, password: targetPassword)
            try await api.applyNow(ssid: targetSSID)
        } catch {
            phase = .failed("Couldn't set WiFi on the drone: \(error.localizedDescription)"); return
        }

        // 4) The drone drops its hotspot to join `targetSSID`; macOS then
        //    auto-rejoins the remembered network. Nudge it if we can see it.
        phase = .waitingForDrone
        if let home = try? await Self.scan(iface).first(where: { $0.ssid == targetSSID }) {
            try? iface.associate(to: home, password: targetPassword)
        }

        // 5) Wait for the drone to reappear on the shared network via Bonjour.
        if let host = await Self.waitForDrone() {
            phase = .done(host)
            onDrone(host)
        } else {
            phase = .failed("Drone didn't come back online — check the WiFi password and retry")
        }
    }

    // MARK: - Helpers

    /// CoreWLAN scan is blocking; run it off the main thread.
    private static func scan(_ iface: CWInterface) async throws -> Set<CWNetwork> {
        try await withCheckedThrowingContinuation { cont in
            DispatchQueue.global(qos: .userInitiated).async {
                do { cont.resume(returning: try iface.scanForNetworks(withSSID: nil)) }
                catch { cont.resume(throwing: error) }
            }
        }
    }

    /// Poll the drone's backend until it answers (or times out).
    private static func waitForBackend(_ api: DroneAPI, timeout: TimeInterval = 25) async -> Bool {
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            if (try? await api.wifiStatus()) != nil { return true }
            try? await Task.sleep(nanoseconds: 1_000_000_000)
        }
        return false
    }

    /// Wait for any drone to appear via Bonjour, returning its host/IP.
    private static func waitForDrone(timeout: TimeInterval = 60) async -> String? {
        await withCheckedContinuation { cont in
            Task { @MainActor in
                var resumed = false
                let browser = BonjourDroneBrowser { ip, _ in
                    guard !resumed else { return }
                    resumed = true
                    cont.resume(returning: ip)
                }
                browser.start()
                DispatchQueue.main.asyncAfter(deadline: .now() + timeout) {
                    browser.stop()                       // keeps browser alive until fired
                    guard !resumed else { return }
                    resumed = true
                    cont.resume(returning: nil)
                }
            }
        }
    }
}
